### ISSUE TITLE
DDG4 simulation with CAD geometry !!?

### ISSUE THREAD

**Issue by Shivang-UHM** [AIDASoft/DD4hep#879]
Hi,

Here are the details of the DD4hep installation

 - OS version: e.g. VBox with Ubuntu 20.04.2 LTS 
 - Compiler version: GCC 9.3
 - DD4hep version:  v01-16-01
 - Reproduced by: Geant4 10.07p01, ROOT version: 6.22, cmake: 3.16.3
 - Input: HMBv3.xml

 - Goal: Geant4 (DDG4) simulation of detector geometry having both CAD structures and DDG4 Box segment
  
So, we have a simple geometry of our Muon Beamline (detector). It consists of 2 tracker planes and a set of 6 calorimeters stacked in the matrix of 2:3. And these tracker planes and calorimeters are exported in STL format from the CAD software Solidworks. We have taken the help from the DDCAD examples and imported the CAD geometries using XML description. 
First, we wanted to record some hits in the calorimeter by showering particles. So, we declared Calorimeter Geometry as sensitive. The compact XML description of geometry is attached (HMBv3.xml). We haven't written our own constructor and hoping existing detector constructors should be good for starting.
After initiating the ddsim simulation or python-based script it shows some initial Geant4 things and then the system kind of freezes. It counts some G4TriangularFacets and sometimes ends with the message that one of the logical volumes is outside the world volume, sometimes segmentation fault and sometimes I need to restart my VBox. 
I have tried to capture some logs please have a look and suggest where am I going wrong.
Could be some fundamental blunder as I'm just a beginner in DD4hep and still trying to understand your wonderful effort (DD4hep).
Any example where you have imported a CAD geometry and ran a Geant4 (DDG4) simulation? 
Attachment: Xml description along with cad files, python script and output log.
[DD4hep_HMB.zip](https://github.com/AIDASoft/DD4hep/files/7464411/DD4hep_HMB.zip)

Looking forward to your kind response.

Thank you,
Shivang Tripathi,
IDLab, UH Manoa

**Comment by MarkusFrankATcernch** [comment]
I tried out your example.
1) The SciFi STL files do not properly translate to Geant4. There must be some errors in them.
2) The calorimeter (6CalAssembly_aluminumMaterial.STL) can be translated to Geant4. 

However, you need some sort of segmentation:
``` 
 <readouts>
    <readout name="CaloHits">
      <segmentation type="CartesianGridXYZ" grid_size_x="15" grid_size_y="15" grid_size_z="15"/>
      <id>system:8,x:20:-8,y:36:-8,z:48:-8</id> 
    </readout>
  </readouts>
```

and you have to supply some material:
```
    <detector id="4" name="6CalAssembly_aluminumMaterial" type="DD4hep_TestShape_Creator" readout="CaloHits" sensitive="true">
      <sensitive type="calorimeter"/>
      <material  name="Silicon"/>
      <check vis="Calo_vis">
        <shape type="CAD_Shape" ref="${DD4hepExamplesINSTALL}/examples/DDHMB/compact/6CalAssembly_aluminumMaterial.STL" unit="mm"/>
          <!--<physvolid name="6CalAssembly_aluminumMaterial" value="1"/> -->
          <position x="0*cm" y="0*cm" z="0"/>
          <rotation x="0" y="0" z="0"/>
      </check>
    </detector>
```
There are apparently still some problematic areas, because Geant4 sometimes gets unhappy:
```
-------- WWWW ------- G4Exception-START -------- WWWW -------
*** G4Exception : GeomNav1002
      issued by : G4Navigator::ComputeStep()
Stuck Track: potential geometry or navigation problem.
  Track stuck, not moving for 10 steps.
  Current  phys volume: 'AV_1!6CalAssembly_aluminumMaterial_0#0!6CalAssembly_aluminumMaterial_vol_0_0#0'
   - at position : (-0.0272340391645529,27.78249931335449,0.06903279087659904)
     in direction: (-0.003834633556570828,-0.9999926285864941,-0.000195852396284496)
    (local position: (-0.0272340391645529,27.78249931335449,0.06903279087659904))
    (local direction: (-0.003834633556570828,-0.9999926285864941,-0.000195852396284496)).
  Previous phys volume: 'world_volume_1'

  Likely geometry overlap - else navigation problem !
   *** Trying to get *unstuck* using a push - expanding step to 1e-07 (mm) ...       Potential overlap in geometry !

*** This is just a warning message. ***
-------- WWWW -------- G4Exception-END --------- WWWW -------
```

The result looks then something like this:
![Screenshot from 2021-11-03 13-58-44](https://user-images.githubusercontent.com/8831087/140065709-ea7f8a95-7d78-474a-a77c-7e5a159416ca.png)

**Comment by Shivang-UHM** [comment]
Hi Markus,

Thanks a lot. I will check the SciFi .stl file and will retry to run the DDG4.
Would like to know more about the segmentation and <id> system:8... what do these numbers represent and how does it works?
And can I define material like NaI for calorimeter instead of having Si?

Thank you

**Comment by MarkusFrankATcernch** [comment]
-- Please see DDDetectors/compact/materials.xml for the construction of "new" composite materials like NaI
-- The segmentations are used define unique readout channels depending on your sensitive volumes. For CAD volumes this is not really applicable. The "system" tag here defines a unique tag for each subdetector system and must have the same size for ALL subdetectors. 8 bits turned out to be appropriate. The other 64-8 bits can then be used to define unique channels within the subdetector.

**Comment by Shivang-UHM** [comment]
Thank you Markus for your help.

**Comment by Shivang-UHM** [comment]
Hi, 

Would like you to please relook into it.

We look for possible flaws in Sci-Fi tracker geometry. Tried to use software inbuilt tools to make sure geometry is fine. Tried to exports in few other formats. But encounter similar error while running DDG4 simulation.
----
When I tried only Calorimeter geometry, I could register some hits as you have shown in your first reply. 
Next, we tried only Sci-Fi tracker geometry and the DDG4 simulation was completed.
So, we have added material, sensitive type info etc. for the tracker geometry but got similar error as we were getting earlier while running ddg4 simulation. 

In short, when there is only 1 CAD geometry in xml file, the DDG4 simulation completes and it registers some hits otherwise fails.
[DD4hep_HMB-2.zip](https://github.com/AIDASoft/DD4hep/files/7708505/DD4hep_HMB-2.zip)
 
Would appreciate if you can have another look.

Thank you,
Shivang

**Comment by MarkusFrankATcernch** [comment]
Hi Shivang,

No there are unfortunately no updates on this issue, because I am currently 100% occupied with LHCb online commissioning.
I expect that the debugging will be quite tedious, hence if this is urgent, it is perhaps better not to count on this feature in the very near future. My apologies for this unfortunate coincidence of needs.

Markus

**Comment by Shivang-UHM** [comment]
Hi Mark,

No worries, I can understand. It is not really that urgent or important at this time.
Thank you

**Comment by MarkusFrankATcernch** [comment]
Hi Shivang,
There is some problem in Geant4 when it comes to attach tessellated shapes to assemblies.
I changed your example by attaching the CAD shapes to a real mother volume (here a box) and then it all works.
Please give it a try. Why exactly this does not work with assemblies I do not really know. But you are correct:
as soon as there is more than one CAD shape attached to an assembly it fails....
I can only advise to place the CAD shapes into real volumes.

[DDHMB.tar.gz](https://github.com/AIDASoft/DD4hep/files/8317493/DDHMB.tar.gz)
[SingleShape_geo.cpp.gz](https://github.com/AIDASoft/DD4hep/files/8317558/SingleShape_geo.cpp.gz)

To run, place the cpp file into examples/ClientTests/src and rebuild these examples.

Cheers,

Markus

**Comment by MarkusFrankATcernch** [comment]
Update: These work better for simulation. Give hits in all detectors
[HMBv3.xml.gz](https://github.com/AIDASoft/DD4hep/files/8317921/HMBv3.xml.gz)
[HMB_Sim.py.gz](https://github.com/AIDASoft/DD4hep/files/8317927/HMB_Sim.py.gz)

**Comment by Shivang-UHM** [comment]
Hi Mark,
I really appreciate you for finding time and coming up with the solution.
It been a while so it took me sometime to get back to the VM where dd4hep is installed
It seems to be working now, creating some hits in all trackers but I could see lot of warning messages (any ideas? should I worry about them?):

```
-------- WWWW ------- G4Exception-START -------- WWWW -------
*** G4Exception : GeomNav1002
      issued by : G4Navigator::ComputeStep()
Stuck Track: potential geometry or navigation problem.
  Track stuck, not moving for 10 steps.
  Current  phys volume: 'SciFiBoard_upper_vol_1'
   - at position : (2.82849626221536,553.3555442115473,-0.5066026809198916)
     in direction: (0.004898661555961355,0.9999874573143328,-0.001043230068320299)
    (local position: (2.82849626221536,3.35554421154734,-0.5066026809198916))
    (local direction: (0.004898661555961355,0.9999874573143328,-0.001043230068320299)).
  Previous phys volume: 'SciFiBoard_upper_vol_0'

  Likely geometry overlap - else navigation problem !
   *** Trying to get *unstuck* using a push - expanding step to 1e-07 (mm) ...       Potential overlap in geometry !

*** This is just a warning message. ***
-------- WWWW -------- G4Exception-END --------- WWWW -------
```
![Screenshot_2022-03-28_19-40-44](https://user-images.githubusercontent.com/60454241/160540967-6f90f4d4-d02d-424a-8efe-4e38fcf85a42.png)

![Screenshot_2022-03-28_19-43-02](https://user-images.githubusercontent.com/60454241/160541228-d2ea5717-ba71-4a83-8597-3ad66086a1fc.png)


I will have to look again in the codes, setting up proper dimensions, etc..
Hope, I can modify the materials for the trackers & calorimeters and continue with DDG4 simulation.

Hoping to update you further on the progress in the coming days.

Thank you,

Shivang

**Comment by MarkusFrankATcernch** [comment]
@Shivang-UHM Can this issue be closed or are there still problems pending?
